### PR TITLE
Gemm/MatMul improvements: support small matrices, stacking and bias broadcasting

### PR DESCRIPTION
**Pull request by pixelspark**
First attempt at fixing #94. 

This one does not work just right yet however because it is still using Mat4x4 for multiplication, which should probably Mat2x2 (when i_shape[1][1] == 2) or even Mat3x3 or Scalar... leading to intermittent blocks of NaNs all over the place in the output:

image

@haixuanTao can you have a look at this, perhaps give me some pointers? I am not too familiar with the Gemm implementation.

### CODE REVIEW

**Comment by pixelspark** [comment]
To illustrate the issue, see the two added test cases in `tests/matrix.rs`. The first one works just fine (4x4 multiplied by 4x4). 

[The second one](https://github.com/webonnx/wonnx/pull/95/files#diff-1a6e1d9c228d79f84f8dba01b6b4068d57fe0671d70ec8ff84a63335b3c484adR207) basically does the following:

````python
a = np.arange(0,16).reshape((4,4))
b = np.ones((4,2))
c = np.matmul(a, b)
// c = array([[ 6.,  6.], [22., 22.], [38., 38.], [54., 54.]])
````

It however currently fails, returning `[1., 1., 1., 1., 9., 9., 9., 9.,]`.

**Comment by haixuanTao** [comment]
Yeah, so I implemented GEMM for very large matrix, that was always divisible by 4. I guess, one way to fix this issue is to introduce a new shader that do scalar / Mat2x2 multiplication as you said. Do you want to have a go at it? Or if you want I can do it :)

**Comment by pixelspark** [comment]
> Yeah, so I implemented GEMM for very large matrix, that was always divisible by 4. I guess, one way to fix this issue is to introduce a new shader that do scalar / Mat2x2 multiplication as you said. Do you want to have a go at it? Or if you want I can do it :)

Hm, I spent half a day trying to make a 'generic' version (where the 'kernel size' can be chosen, e.g. 2x2 or 4x4) but somehow it doesn't work quite yet (and I just don't see it anymore 😵 ). @haixuanTao would really appreciate if you could have a look (latest attempt is in this PR but feel free to change it)

File: wonnx/templates/matrix/gemm.wgsl [hunk 1]
```diff
@@ -1,71 +1,102 @@
...
+	[[group(0), binding(2)]]
+	var<storage, write> output_0: GemmArrayVector;
+{% endif %}
 
 [[stage(compute), workgroup_size(1)]]
 fn main([[builtin(global_invocation_id)]] global_id: vec3<u32>) {
-	let y = global_id.x % {{ i_shape[1][1] / 4 | int }}u;
-	let x = global_id.x / {{ i_shape[1][1] / 4 | int }}u;
+	{% if i_shape[0][1] < kernel_size %}
+		let n_k = 1u;
```
Comment (haixuanTao) [inline wonnx/templates/matrix/gemm.wgsl:29]: n could refer to too many things at this point. Having a more distinctive name like: `output_rows` for n, and explicating that you're doing a division might help. `output_rows_div_kernel`, is pretty verbose, but a lot easier to get than `n_k`

**Comment by pixelspark** [comment]
OK, I think I figured it out, at least all test cases are passing now. Hopefully the code is also more understandable to you now as well (in the code I refer to the M, N and K dimensions as also used [here](https://github.com/onnx/onnx/blob/main/docs/Operators.md#gemm). Dimension constraints: `(M*K)*(K*N) = (M*N)`). 

The code now dynamically selects the appropriate kernel size (4 if M/N/K are divisible by 4, 2 if M/N/K are divisible by 2 and not by 4).  Note this still doesn't work for e.g. `(3x3)*(3x2)=(3x2)` so I made a note in the README and explicitly return an error for these cases. I also added in the `workgroup_size` determination so we can now actually multiply larger matrixes as well without running into the limits to the number of (x-)threads.

**Comment by pixelspark** [comment]
Note that the official test cases for Gemm fail even with this change:

````python
backend_test.include(f"test_gemm_")
````

Output for this PR:
````
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_all_attributes_cpu - pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: GpuError(CompileError(InvalidInputShape { input_index: 0, input_shape: Shape { dims: [4, 3], data_type: F32 } }))
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_alpha_cpu - pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: GpuError(CompileError(UnimplementedVariant { variant: "with the greatest common divider of dimensions M, N or K not divisible by 2", op: "Gemm" }))
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_beta_cpu - pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: GpuError(CompileError(UnimplementedVariant { variant: "with the greatest common divider of dimensions M, N or K not divisible by 2", op: "Gemm" }))
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_default_matrix_bias_cpu - pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: GpuError(CompileError(UnimplementedVariant { variant: "with the greatest common divider of dimensions M, N or K not divisible by 2", op: "Gemm" }))
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_default_no_bias_cpu - pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: GpuError(CompileError(UnimplementedVariant { variant: "with the greatest common divider of dimensions M, N or K not divisible by 2", op: "Gemm" }))
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_default_scalar_bias_cpu - pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: GpuError(CompileError(UnimplementedVariant { variant: "with the greatest common divider of dimensions M, N or K not divisible by 2", op: "Gemm" }))
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_default_single_elem_vector_bias_cpu - pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: GpuError(CompileError(UnimplementedVariant { variant: "with the greatest common divider of dimensions M, N or K not divisible by 2", op: "Gemm" }))
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_default_vector_bias_cpu - pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: GpuError(CompileError(UnimplementedVariant { variant: "with the greatest common divider of dimensions M, N or K not divisible by 2", op: "Gemm" }))
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_default_zero_bias_cpu - pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: GpuError(CompileError(UnimplementedVariant { variant: "with the greatest common divider of dimensions M, N or K not divisible by 2", op: "Gemm" }))
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_transposeA_cpu - pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: GpuError(CompileError(InvalidInputShape { input_index: 0, input_shape: Shape { dims: [6, 3], data_type: F32 } }))
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_transposeB_cpu - pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: GpuError(CompileError(InvalidInputShape { input_index: 1, input_shape: Shape { dims: [4, 6], data_type: F32 } }))
````

Master (also fails):
````
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_all_attributes_cpu - pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: GpuError(CompileError(UnimplementedVariant { variant: "Gemm with transA/transB/broadcast not equal to zero", op: "Gemm" }))
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_alpha_cpu - AssertionError: 
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_beta_cpu - AssertionError: 
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_default_matrix_bias_cpu - AssertionError: 
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_default_no_bias_cpu - AssertionError: 
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_default_scalar_bias_cpu - AssertionError: 
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_default_single_elem_vector_bias_cpu - AssertionError: 
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_default_vector_bias_cpu - AssertionError: 
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_default_zero_bias_cpu - AssertionError: 
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_transposeA_cpu - pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: GpuError(CompileError(UnimplementedVariant { variant: "Gemm with transA/transB/broadcast not equal to zero", op: "Gemm" }))
FAILED tests/test_onnx_backend.py::OnnxBackendNodeModelTest::test_gemm_transposeB_cpu - pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: GpuError(CompileError(UnimplementedVariant { variant: "Gemm with transA/transB/broadcast not equal to zero", op: "Gemm" }))
````

At least this PR gives a bit more descriptive errors... :-)

**Comment by haixuanTao** [comment]
I'm sorry but using M, N, K as abbreviation is not ok, even though in the context of developing GEMM application it might seems alright. 
`with the greatest common divider of dimensions M, N or K not divisible by 2", op: "Gemm"`
This is a lot harder to debug than:
`Input matrix dimensions are not divisible by 2 which is required for the current Matrix Kernel division", op: "Gemm"`

**Comment by pixelspark** [comment]
OK, sure, I changed the error:

````rust
return Err(CompileError::UnimplementedVariant {
                        variant: String::from(
                            "usage with input matrixes whose dimensions are not divisible by 2",
                        ),
                        op: op.to_string(),
                    });
````

This will print as:

````
'usage with input matrixes whose dimensions are not divisible by 2' is not yet implemented for op Gemm
````

From my perspective it's fine to keep the references to M/K/N in the code itself (I actually like the fact that these correspond to what the spec calls them, but then I guess we do not necessarily expect readers of the code to know the spec). If you prefer I could rename them (i.e. `dims_m` could become `output_height` or `output_rows` or similar).

**Comment by pixelspark** [comment]
OK, so I implemented broadcasting for bias as well, appears to work (will add test cases later but output for SQuAD and MNIST now looks as expected).

I also merged the stacking into this PR so this is one big PR for all Gemm/MatMul related changes.

**Comment by pixelspark** [comment]
Test cases confirm matrix bias and broadcasting bias now working as expected. Still need to fix scalar and vertical broadcasting (test cases written but fail currently).

**Comment by pixelspark** [comment]
@haixuanTao this now passes all the old test cases as well as the new ones I added (based on my interpretation of the ONNX spec, which mainly refers to Numpy, hence why there is numpy code in the documentation of these test cases). 

Let me know what you think!

**Comment by haixuanTao** [comment]
This looks great to me! Thanks for the documentation.

**Pull request merged**